### ISSUE TITLE
Discuss: merge customers_info table into customers table

### ISSUE THREAD

**Issue by drbyte** [zencart/zencart#3847]
**Proposal:** Collapse the fields from the `customers_info` table into the `customers` table.

**Pros:**
- simpler db queries (no need to constantly do joins)
- fewer errors when those joins fail due to missing data
- less data corruption by tools that forget to create corresponding records in this table when adding customers from external means
- slightly more intuitive, as all the info is in one table
- opportunity to shorten the very long field names associated with that table's data

**Cons:**
- several queries to rewrite, thus touching a bunch of core code
- small hit to performance, insomuch as every time a customer logs in their record will be updated, requiring both a lock and a write, which also invalidates MySQL's internal cache for that record. Minor, but real. But actual "login" traffic is very different from regular shopping/browsing traffic, so impact is minimal.
- external tools/contribs that formerly expected that table will need updating
- upgrade SQL will need to merge the data (it's a simple query, but will take a few seconds on large databases; also must accommodate missing records)

**Comment by scottcwilson** [comment]
Could some of the pros also be achieved by creating a customer object?

**Comment by wiztechinc** [comment]
All for simplifying always but not caring one way or the other if this gets done. Meh?

**Comment by drbyte** [comment]
> Could some of the pros also be achieved by creating a customer object?

Yes. I've been working on exactly that, and it reminded me of the limited value of the separate table.

**Comment by lat9** [comment]
I've been wondering for a couple of years why that `customers_info` was a separate table.  Although the change will affect a couple of my plugins ([One-Page Checkout](https://vinosdefrutastropicales.com/index.php?main_page=product_info&cPath=2_7&products_id=69), [Remember Me](https://vinosdefrutastropicales.com/index.php?main_page=product_info&cPath=2_7&products_id=39) and [Favorites Wishlist](https://vinosdefrutastropicales.com/index.php?main_page=product_info&cPath=2_8&products_id=56)), I'm supportive of this change.

**Comment by scottcwilson** [comment]
> I've been wondering for a couple of years why that `customers_info` was a separate table. 

It's like Stonehenge.  It was done by our ancestors for reasons we cannot fathom.

**Comment by vandijkstef** [comment]
Im all for separating queries and (page-)logic by extracting them into objects/classes. Files that contain HTML should not contain the actual query IMO. Setting up an object like this will allow third party to use an API layer instead of poking on the database directly, which is the root cause of having third party not properly updating all relevant tables. We can also shorten the column names when using the data. 

I recently did something comparable for 'order_status' and 'order_status_email' on a (heavily modified) 1.5.6 store and was able to remove (extract) just over 200 lines from '/admin/orders.php'. (Let me know if theres interest in adopting this pattern, Ill work on a PR)

`"FROM " . TABLE_CUSTOMERS . "` is found 54 times in 28 files.

Code example (mildly pseudo'd)

``` PHP
// Suggestion (assuming static class/object)
$customerData = $customer->get([id, name]) // $fields: string | string[] = '*'
$id = $customerData['id']

// Current
$customerSQL = "SELECT customers_id, customers_name, customers_something, customers_customers FROM i_dont_care WHERE 'lookup_how_its_been_done_elsewhere";
$customerData = $db->Execute($customerSQL);
$id = $customerData['customers_id'];
```

**TLDR: Up for simplifying, but allow third party to not fuss about the actual database and/or changes on the database first. (We're not killing any fields, beside from where they live, field names remain 'date_of_last_logon' in code) **
